code/game/objects/items/RCD.dm

```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
 	GLOB.rcd_list -= src
 	. = ..()
 
-/obj/item/construction/rcd/attack_self(mob/user)
-	..()
-	var/list/choices = list(
-		"Airlock" = image(icon = 'icons/hud/radial.dmi', icon_state = "airlock"),
-		"Grilles & Windows" = image(icon = 'icons/hud/radial.dmi', icon_state = "grillewindow"),
-		"Floors & Walls" = image(icon = 'icons/hud/radial.dmi', icon_state = "wallfloor")
+
+/obj/item/construction/rcd/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/rcd),
 	)
-	if(upgrade & RCD_UPGRADE_FRAMES)
-		choices += list(
-		"Machine Frames" = image(icon = 'icons/hud/radial.dmi', icon_state = "machine"),
-		"Computer Frames" = image(icon = 'icons/hud/radial.dmi', icon_state = "computer_dir"),
-		)
-	if(upgrade & RCD_UPGRADE_SILO_LINK)
-		choices += list(
-		"Silo Link" = image(icon = 'icons/obj/mining.dmi', icon_state = "silo"),
-		)
-	if(upgrade & RCD_UPGRADE_FURNISHING)
-		choices += list(
-		"Furnishing" = image(icon = 'icons/hud/radial.dmi', icon_state = "chair")
-		)
-	switch(construction_mode)
-		if(RCD_AIRLOCK)
-			choices += list(
-			"Change Access" = image(icon = 'icons/hud/radial.dmi', icon_state = "access"),
-			"Change Airlock Type" = image(icon = 'icons/hud/radial.dmi', icon_state = "airlocktype")
-			)
-		if(RCD_WINDOWGRILLE)
-			choices += list(
-			"Change Window Glass" = image(icon = 'icons/hud/radial.dmi', icon_state = "windowtype"),
-			"Change Window Size" = image(icon = 'icons/hud/radial.dmi', icon_state = "windowsize")
-			)
-		if(RCD_FURNISHING)
-			choices += list(
-			"Change Furnishing Type" = image(icon = 'icons/hud/radial.dmi', icon_state = "chair")
-			)
-	var/choice = show_radial_menu(user, src, choices, custom_check = CALLBACK(src, PROC_REF(check_menu), user), require_near = TRUE, tooltips = TRUE)
-	if(!check_menu(user))
+
+/obj/item/construction/rcd/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "RapidConstructionDevice", name)
+		ui.open()
+
+/**
+ * if ui_always_active = TRUE display & update window even if nothing changed, required for construction_console else window wont show up
+ * else use parent method to decide the state for normal usage
+ */
+/obj/item/construction/rcd/ui_state(mob/user)
+	return ui_always_active ? GLOB.always_state : ..()
+
+/obj/item/construction/rcd/ui_static_data(mob/user)
+	return airlock_electronics.ui_static_data(user)
+
+/obj/item/construction/rcd/ui_data(mob/user)
+	var/list/data = ..(user)
+
+	data["space"] = TRUE
```

Comment:
This shouldn't be done through server side. Why is it like this?

Reply:
Just learnt how to use props in typescript.  I'll change it noow